compiler/rustc_hir_typeck/src/demand.rs

```diff
@@ -693,6 +694,55 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
         );
     }
 
+    pub(crate) fn note_result_coercion(
+        &self,
+        err: &mut Diagnostic,
+        expr: &hir::Expr<'tcx>,
+        expected: Ty<'tcx>,
+        found: Ty<'tcx>,
+    ) -> bool {
+        let ty::Adt(e, substs_e) = expected.kind() else { return false; };
+        let ty::Adt(f, substs_f) = found.kind() else { return false; };
+        if e.did() != f.did() {
+            return false;
+        }
+        if Some(e.did()) != self.tcx.get_diagnostic_item(sym::Result) {
+            return false;
+        }
+        let map = self.tcx.hir();
+        if let Some(hir::Node::Expr(expr)) = map.find_parent(expr.hir_id)
+            && let hir::ExprKind::Ret(_) = expr.kind
+        {
+            // `return foo;`
+        } else if map.get_return_block(expr.hir_id).is_some() {
+            // Function's tail expression.
+        } else {
+            return false;
+        }
+        let e = substs_e.type_at(1);
+        let f = substs_f.type_at(1);
+        if self
+            .infcx
+            .type_implements_trait(
+                self.tcx.get_diagnostic_item(sym::Into).unwrap(),
+                [f, e],
+                self.param_env,
+            )
+            .must_apply_modulo_regions()
+        {
+            err.multipart_suggestion(
+                "you can rely on the implicit conversion that `?` does to transform the error type",
```

Comment:
Maybe something like

```suggestion
                "use `?` then wrap the value in `Ok` to return an `Err` result and coerce the `Ok` variant into the right `Result` type",
```

?? no idea if that's clearer 😅

Reply:
I'm not too happy with that explanation because I feel it is too much. I want us to take into consideration the context we are in: we have some `Result<_, A>` and we expected `Result<_, B>` where `B: Into<A>`. Because of that I want to focus on "writing the following you can turn the value you have into the expected type". Maybe "you can use `?` to coerce return an appropriate `Err`, and wrap the resulting value in `Ok` so it remains a `Result`"?

Reply:
I'm fine with that wording minus the "you can". I personally prefer suggestions worded in the imperative mood, but r=me either way.